ethcore/src/client/client.rs

```diff
@@ -417,21 +418,48 @@ impl<V> BlockChainClient for Client<V> where V: Verifier {
 
 		// Add transactions
 		let block_number = b.block().header().number();
+		let min_tx_gas = U256::from(self.engine.schedule(&b.env_info()).tx_gas);
+		let gas_limit = *b.block().header().gas_limit();
+		let mut gas_left = gas_limit;
+		let mut invalid_transactions = HashSet::new();
+
 		for tx in transactions {
+			// Stop early if we are sure that no other transaction will be included
+			if gas_left < min_tx_gas {
+				break;
+			}
+
+			// TODO [todr] It seems that calculating gas_left here doesn't look nice. After moving this function
+			// to miner crate we should consider rewriting this logic in some better way.
+			if tx.gas > gas_left {
```

Comment:
Right, I guess it would be possible to unwrap error, check if it's `BlockGasLimitReached`, check minimal transaction gas.

Will refactor, because calculating gas here is not particularly appealing.